packages/zoe/src/contractFacet/contractFacet.js

```diff
@@ -0,0 +1,348 @@
+          // verifies offer safety
+          const seatStaging = zcfSeat.stage(newAllocation);
+          // No effects above. Commit point. The following two steps
+          // *should* be committed atomically.
+          // But unlike https://github.com/Agoric/agoric-sdk/issues/1391
+          // it is not a disater if they are not.
+          // If we minted only, no one would ever get those
+          // invisibly-minted assets.
+          E(zoeMintP).mintAndEscrow(totalToMint);
+          zcfSeatAdmin.commit(seatStaging);
+          return zcfSeat;
+        },
+        burnLosses: (losses, zcfSeat) => {
+          let totalToBurn = mintyAmountMath.getEmpty();
+          const oldAllocation = zcfSeat.getCurrentAllocation();
+          const updates = objectMap(
+            losses,
+            ([seatKeyword, amountToSubtract]) => {
+              totalToBurn = mintyAmountMath.add(totalToBurn, amountToSubtract);
+              const oldAmount = oldAllocation[seatKeyword];
+              const newAmount = mintyAmountMath.subtract(
+                oldAmount,
+                amountToSubtract,
+              );
+              return [seatKeyword, newAmount];
+            },
+          );
+          const newAllocation = harden({
+            ...oldAllocation,
+            ...updates,
+          });
+          const zcfSeatAdmin = seatToZCFSeatAdmin.get(zcfSeat);
+          // verifies offer safety
+          const seatStaging = zcfSeat.stage(newAllocation);
+          // No effects above. Commit point. The following two steps
+          // *should* be committed atomically.
+          // But unlike https://github.com/Agoric/agoric-sdk/issues/1391
+          // it is not a disater if they are not.
+          // If we only commit the staging, no one would ever get the
+          // unburned assets.
+          zcfSeatAdmin.commit(seatStaging);
+          E(zoeMintP).withdrawAndBurn(totalToBurn);
+        },
+      });
+      return zcfMint;
+    };
+
+    /** @type ContractFacet */
+    const zcf = {
+      reallocate: (/** @type SeatStaging[] */ ...seatStagings) => {
+        // We may want to handle this with static checking instead.
+        // Discussion at: https://github.com/Agoric/agoric-sdk/issues/1017
+        assert(
+          seatStagings.length >= 2,
+          details`reallocating must be done over two or more seats`,
+        );
+
+        seatStagings.forEach(seatStaging =>
+          assert(
+            allSeatStagings.has(seatStaging),
+            details`The seatStaging ${seatStaging} was not recognized`,
+          ),
+        );
+
+        // Ensure that rights are conserved overall. Offer safety was
+        // already checked when an allocation was staged for an individual seat.
+        const flattened = arr => [].concat(...arr);
+        const flattenAllocations = allocations =>
+          flattened(allocations.map(allocation => Object.values(allocation)));
```

Comment:
`flatMap` is now an official part of EcmaScript
https://tc39.es/ecma262/#sec-array.prototype.flatmap
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap
I think it already ships on all the engines we care about.

```suggestion
        // Ensure that rights are conserved overall. Offer safety was
        // already checked when an allocation was staged for an individual seat.
        const flattenAllocations = allocations => 
          allocations.flatMap(Object.values);
```

Reply:
Huh, I had tried `prototype.flat` which did not work but will look into `flatMap` https://github.com/Agoric/agoric-sdk/issues/1157

Reply:
also called "smoosh" :D